File: cxx_wrapper/FragmentAssembler.cpp

#include "rust/cxx.h"
#include <memory>
#include "FragmentAssembler.h"

namespace aeron { namespace fragment_assembler {

inline std::unique_ptr<aeron::FragmentAssembler> newInstance(const rust::Fn<void(aeron::concurrent::AtomicBuffer const &, std::int32_t, std::int32_t, aeron::concurrent::logbuffer::Header const &)> fragmentHandler, std::size_t initialBufferLength) {
     return std::unique_ptr<aeron::FragmentAssembler>(new aeron::FragmentAssembler(fragmentHandler, initialBufferLength));
}

inline  void deleteSessionBuffer(const std::unique_ptr<aeron::FragmentAssembler> &fragment_assembler, std::int32_t sessionId){
    fragment_assembler->deleteSessionBuffer(sessionId);
}

}}

File: src/aeron/fragment_assembler.rs

#[cxx::bridge(namespace = "aeron")]
pub mod ffi {

    unsafe extern "C++" {
        #[namespace = "aeron::concurrent"]
        #[rust_name = "CxxAtomicBuffer"]
        type AtomicBuffer = crate::aeron::concurrent::atomic_buffer::ffi::CxxAtomicBuffer;

        #[namespace = "aeron::concurrent::logbuffer"]
        #[rust_name = "CxxHeader"]
        type Header = crate::aeron::concurrent::logbuffer::header::ffi::CxxHeader;
        #[namespace = "aeron"]
        #[rust_name = "CxxControlledPollAction"]
        type ControlledPollAction = crate::aeron::image::ffi::CxxControlledPollAction;

        #[namespace = "aeron"]
        #[rust_name = "CxxImage"]
        type Image = crate::aeron::image::ffi::CxxImage;

        include!("aeron-rust-wrapper/aeron/aeron-client/src/main/cpp/FragmentAssembler.h");
        #[rust_name = "CxxFragmentAssembler"]
        type FragmentAssembler;


        include!("aeron-rust-wrapper/cxx_wrapper/FragmentAssembler.cpp");


        //fn handler(self: &CxxFragmentAssembler) -> fn(&CxxAtomicBuffer, i32, i32, &CxxHeader) -> ();

        #[namespace = "aeron::fragment_assembler"]
        fn deleteSessionBuffer(fragment_assembler: &UniquePtr<CxxFragmentAssembler>, sessionId: i32);


        #[namespace = "aeron::fragment_assembler"]
        fn newInstance(handler: fn(&CxxAtomicBuffer, i32, i32, &CxxHeader) -> (), initialBufferLength: usize) -> UniquePtr<CxxFragmentAssembler>;
    }

}